core/net/rpl/rpl-icmp6.c

```diff
@@ -721,84 +783,121 @@ dao_input(void)
-                              0, NBR_REACHABLE)) != NULL) {
-      /* set reachable timer */
-      stimer_set(&nbr->reachable, UIP_ND6_REACHABLE_TIME / 1000);
-      PRINTF("RPL: Neighbor added to neighbor cache ");
-      PRINT6ADDR(&dao_sender_addr);
-      PRINTF(", ");
-      PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
-      PRINTF("\n");
-    } else {
-      PRINTF("RPL: Out of Memory, dropping DAO from ");
-      PRINT6ADDR(&dao_sender_addr);
-      PRINTF(", ");
-      PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
-      PRINTF("\n");
-      goto discard;
+  /* Update and add neighbor - if no room - fail. */
+  if((nbr = rpl_icmp6_update_nbr_table(&dao_sender_addr, NBR_TABLE_REASON_RPL_DAO, instance)) == NULL) {
+    PRINTF("RPL: Out of Memory, dropping DAO from ");
+    PRINT6ADDR(&dao_sender_addr);
+    PRINTF(", ");
+    PRINTLLADDR((uip_lladdr_t *)packetbuf_addr(PACKETBUF_ADDR_SENDER));
+    PRINTF("\n");
+    if(flags & RPL_DAO_K_FLAG) {
+      /* signal the failure to add the node */
+      dao_ack_output(instance, &dao_sender_addr, sequence,
+		     is_root ? RPL_DAO_ACK_UNABLE_TO_ACCEPT_ROOT :
+		     RPL_DAO_ACK_UNABLE_TO_ACCEPT);
     }
-  } else {
-    PRINTF("RPL: Neighbor already in neighbor cache\n");
+    goto discard;
   }
 
   rep = rpl_add_route(dag, &prefix, prefixlen, &dao_sender_addr);
   if(rep == NULL) {
     RPL_STAT(rpl_stats.mem_overflows++);
     PRINTF("RPL: Could not add a route after receiving a DAO\n");
+    if(flags & RPL_DAO_K_FLAG) {
+      /* signal the failure to add the node */
+      dao_ack_output(instance, &dao_sender_addr, sequence,
+		     is_root ? RPL_DAO_ACK_UNABLE_TO_ACCEPT_ROOT :
+		     RPL_DAO_ACK_UNABLE_TO_ACCEPT);
+    }
     goto discard;
   }
 
+  /* set lifetime and clear NOPATH bit */
   rep->state.lifetime = RPL_LIFETIME(instance, lifetime);
-  rep->state.learned_from = learned_from;
-  rep->state.nopath_received = 0;
+  RPL_ROUTE_CLEAR_NOPATH_RECEIVED(rep);
 
 #if RPL_CONF_MULTICAST
 fwd_dao:
 #endif
 
   if(learned_from == RPL_ROUTE_FROM_UNICAST_DAO) {
+    int should_ack = 0;
+
+    if(flags & RPL_DAO_K_FLAG) {
+      /* check if this route is already installed and we can ack now! */
+      /* not pending - and same seq-no means that we can ack.         */
+      if((!RPL_ROUTE_IS_DAO_PENDING(rep) &&
+          rep->state.dao_seqno_in == sequence) ||
+          dag->rank == ROOT_RANK(instance)) {
+        should_ack = 1;
+      }
+    }
+
     if(dag->preferred_parent != NULL &&
        rpl_get_parent_ipaddr(dag->preferred_parent) != NULL) {
+      uint8_t out_seq = 0;
+      /* if this is pending and we get the same seq no it is a retrans */
+      if(RPL_ROUTE_IS_DAO_PENDING(rep) &&
+         rep->state.dao_seqno_in == sequence) {
+        /* keep the same seq-no as before for parent also */
+        out_seq = rep->state.dao_seqno_out;
+      } else {
+        out_seq = prepare_for_dao_fwd(sequence, rep);
```

Comment:
That is correct. Not sure if I would like to add more state to handle multiple pending DAOs? I am kind of assuming that the latest is the one to handle - but it might be a too simple assumption?